services/keycloak/start.sh

```diff
@@ -1691,6 +1691,50 @@ function update_openshift_view_permission {
+    echo Creating group lagoon project IDs mapper for service-api
+    echo '{"protocol":"openid-connect","config":{"id.token.claim":"false","access.token.claim":"true","userinfo.token.claim":"false","multivalued":"true","aggregate.attrs":"true","user.attribute":"group-lagoon-project-ids","claim.name":"group_lagoon_project_ids","jsonType.label":"String"},"name":"Group Lagoon Project IDs","protocolMapper":"oidc-usermodel-attribute-mapper"}' | /opt/jboss/keycloak/bin/kcadm.sh create -r ${KEYCLOAK_REALM:-master} clients/$CLIENT_ID/protocol-mappers/models --config $CONFIG_PATH -f -
+}
+
+function configure_token_exchange {
+    REALM_MANAGEMENT_CLIENT_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get  -r lagoon clients?clientId=realm-management --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+    IMPERSONATE_PERMISSION_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get -r lagoon clients/$REALM_MANAGEMENT_CLIENT_ID/authz/resource-server/permission?name=admin-impersonating.permission.users --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+    # check if policies already exist
+    service_api_client_policy_id=$(/opt/jboss/keycloak/bin/kcadm.sh get -r lagoon clients/$REALM_MANAGEMENT_CLIENT_ID/authz/resource-server/policy?name=Client+service-api+Policy --config $CONFIG_PATH)
+    if [ "$service_api_client_policy_id" != "[ ]" ]; then
+        echo "Client service-api policies are already created, skipping policy setup"
+        return 0
+    fi
+
+    # create client service-api policy
+    SERVICE_API_CLIENT_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get  -r lagoon clients?clientId=service-api --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+    echo '{"type":"client","logic":"POSITIVE","decisionStrategy":"UNANIMOUS","name":"Client service-api Policy","clients":["'$SERVICE_API_CLIENT_ID'"]}' | /opt/jboss/keycloak/bin/kcadm.sh create clients/$REALM_MANAGEMENT_CLIENT_ID/authz/resource-server/policy/client --config $CONFIG_PATH -r lagoon -f -
+    SERVICE_API_CLIENT_POLICY_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get -r lagoon clients/$REALM_MANAGEMENT_CLIENT_ID/authz/resource-server/policy?name=Client+service-api+Policy --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+
+    echo Enable token exchange for auth-server and service-api
+    # The decision strategy is affirmative since only one policy has to pass.
+    # NOTE: this assumes that configure_api_client has set the global AUTH_SERVER_CLIENT_POLICY_ID variable.
```

Comment:
I'm afraid that assumption will cause problems. There are 3 scenarios each function has to account for:

1. A fresh install ✅: `configure_api_client` will see that it needs to get run and set `AUTH_SERVER_CLIENT_POLICY_ID` needed in this function
2. An already fully configured keycloak ✅: `configure_api_client` [will not run](https://github.com/uselagoon/lagoon/blob/main/services/keycloak/start.sh#L100) and will not set `AUTH_SERVER_CLIENT_POLICY_ID` but `configure_token_exchange` [also won't run](https://github.com/uselagoon/lagoon/pull/2941/files#diff-8a14c757e38a4f4906e6dd229db8406de31fec05d5b8c5e1c6b0514e5a8c1f19R1723) so there will be no issues
3. An upgrade from lagoon <= 2.4.1 ❌: `configure_api_client` [will not run](https://github.com/uselagoon/lagoon/blob/main/services/keycloak/start.sh#L100) and will not set `AUTH_SERVER_CLIENT_POLICY_ID` but `configure_token_exchange` will see that it needs to create the client policy and will probably set the impersonate permission so that it has a null/empty value. Might throw an error? Might fail silently?

You can just copy/paste the `AUTH_SERVER_CLIENT_POLICY_ID` querying code in here so that this function has no dependencies.

Reply:
Oh nicely spotted, thanks!

I'll fix this up as you suggest.

Reply:
Okay, made the change. Does that look better?

Reply:
LGTM!